<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ mesa</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mesa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="BSgenome.Hsapiens.NCBI.GRCh38.CpG.distribution.html">BSgenome.Hsapiens.NCBI.GRCh38.CpG.distribution</a></code> </p>
        </td>
        <td><p>Pre-calculated CpG distribution values for GRCh38</p></td>
      </tr><tr><td>
          <p><code><a href="FantomRegions.html">FantomRegions</a></code> </p>
        </td>
        <td><p>FANTOM5 regions for GRCh38</p></td>
      </tr><tr><td>
          <p><code><a href="GRCh38.99.gtf.html">GRCh38.99.gtf</a></code> </p>
        </td>
        <td><p>GTF file for GRCh38.</p></td>
      </tr><tr><td>
          <p><code><a href="addBamCoveragePairedAndUnpaired.html">addBamCoveragePairedAndUnpaired()</a></code> </p>
        </td>
        <td><p>This function adds coverage to the qseaSet, using both paired reads and the high-quality R1s.</p></td>
      </tr><tr><td>
          <p><code><a href="addHMMcopyCNV.html">addHMMcopyCNV()</a></code> </p>
        </td>
        <td><p>Calculate and add CNV based on HMMcopy, on individual samples</p></td>
      </tr><tr><td>
          <p><code><a href="addHyperStableFraction.html">addHyperStableFraction()</a></code> </p>
        </td>
        <td><p>This function calculates and summarises the proportion of hyperstable regions that are present in each sample, and adds to columns of the sampleTable</p></td>
      </tr><tr><td>
          <p><code><a href="addLibraryInformation.html">addLibraryInformation()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and adds the information on the number of reads into the sampleTable for easier access.</p></td>
      </tr><tr><td>
          <p><code><a href="addMedipsEnrichmentFactors.html">addMedipsEnrichmentFactors()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and adds the Medips-style enrichment scores to the object sampleTable.</p></td>
      </tr><tr><td>
          <p><code><a href="addQseaNormalisationSteps.html">addQseaNormalisationSteps()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and adds several normalisation steps</p></td>
      </tr><tr><td>
          <p><code><a href="annotateData.html">annotateData()</a></code> </p>
        </td>
        <td><p>Add annotation onto a data table</p></td>
      </tr><tr><td>
          <p><code><a href="arrange.qseaSet.html">arrange(<i>&lt;qseaSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>This function extends the function dplyr::arrange to act on qseaSet, to reorder the samples in the qseaSet.</p></td>
      </tr><tr><td>
          <p><code><a href="arrangeQset.html">arrangeQset()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and sorts the samples inside it using dplyr::arrange, using a column(s) of the sample table. Can also use desc() to reverse sort, as in dplyr::arrange.</p></td>
      </tr><tr><td>
          <p><code><a href="calculateArrayBetas.html">calculateArrayBetas()</a></code> </p>
        </td>
        <td><p>This function calculates the beta values for the windows covering the hg38 array probes.</p></td>
      </tr><tr><td>
          <p><code><a href="calculateCpGEnrichment.html">calculateCpGEnrichment()</a></code> </p>
        </td>
        <td><p>This function takes a path to a bam file and calculates the enrichment scores.</p></td>
      </tr><tr><td>
          <p><code><a href="calculateCpGEnrichmentGRanges.html">calculateCpGEnrichmentGRanges()</a></code> </p>
        </td>
        <td><p>This function takes a GRanges object and calculates the enrichment scores.</p></td>
      </tr><tr><td>
          <p><code><a href="calculateDMRs.html">calculateDMRs()</a></code> </p>
        </td>
        <td><p>Fit the GLM and return the data</p></td>
      </tr><tr><td>
          <p><code><a href="calculateFractionReadsInGRanges.html">calculateFractionReadsInGRanges()</a></code> </p>
        </td>
        <td><p>This function calculates the fraction of reads in a qseaSet object which are present in a GRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="calculateGenomicCGDistribution.html">calculateGenomicCGDistribution()</a></code> </p>
        </td>
        <td><p>This function takes a BSGenome and returns a data frame with the relH and GoGe parameters used for calculating enrichment</p></td>
      </tr><tr><td>
          <p><code><a href="colnames.qseaSet.html">colnames(<i>&lt;qseaSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>This function extends the base function length to return the number of samples in the qseaSet</p></td>
      </tr><tr><td>
          <p><code><a href="combineQsets.html">combineQsets()</a></code> </p>
        </td>
        <td><p>Combine two qseaSets</p></td>
      </tr><tr><td>
          <p><code><a href="combineQsetsList.html">combineQsetsList()</a></code> </p>
        </td>
        <td><p>Combine two qseaSets</p></td>
      </tr><tr><td>
          <p><code><a href="countWindowsOverCutoff.html">countWindowsOverCutoff()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and sums over nrpm within a GRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="downsampleQsea.html">downsampleQsea()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and downsamples the reads.</p></td>
      </tr><tr><td>
          <p><code><a href="dropAvgFragDetails.html">dropAvgFragDetails()</a></code> </p>
        </td>
        <td><p>This function removes the avgFragmentLength and avgFragmentMAPQ columns from the regions of the qseaSet</p></td>
      </tr><tr><td>
          <p><code><a href="dropPooledControl.html">dropPooledControl()</a></code> </p>
        </td>
        <td><p>This function drops the PooledControl sample from the qseaSet, if it is present.</p></td>
      </tr><tr><td>
          <p><code><a href="encodeBlacklist.html">encodeBlacklist</a></code> </p>
        </td>
        <td><p>2019 ENCODE blacklist of poorly mapped regions in hg38</p></td>
      </tr><tr><td>
          <p><code><a href="examplePairedTumourQset.html">examplePairedTumourQset</a></code> </p>
        </td>
        <td><p>A small example qseaSet with 5 paired tumour/normal samples.</p></td>
      </tr><tr><td>
          <p><code><a href="filter.qseaSet.html">filter(<i>&lt;qseaSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>This function extends the dplyr function filter to act on qseaSet sampleTable.</p></td>
      </tr><tr><td>
          <p><code><a href="filterByNonOverlaps.html">filterByNonOverlaps()</a></code> </p>
        </td>
        <td><p>Filter a qseaSet by non-overlaps with a GRanges object or table that can be coerced to one</p></td>
      </tr><tr><td>
          <p><code><a href="filterByOverlaps.html">filterByOverlaps()</a></code> </p>
        </td>
        <td><p>Filter a qseaSet by overlaps with a GRanges object or table that can be coerced to one</p></td>
      </tr><tr><td>
          <p><code><a href="filterQset.html">filterQset()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and filters the samples in it based on a call to dplyr::filter on the sampleTable.</p></td>
      </tr><tr><td>
          <p><code><a href="filterRegions.html">filterRegions()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet object and filters the regions inside it by a call to dplyr::filter.</p></td>
      </tr><tr><td>
          <p><code><a href="fitQseaGLM.html">fitQseaGLM()</a></code> </p>
        </td>
        <td><p>Fit a generalised linear model</p></td>
      </tr><tr><td>
          <p><code><a href="gc_hg38_1000kb.html">gc_hg38_1000kb</a></code> <code><a href="gc_hg38_1000kb.html">gc_hg38_10kb</a></code> <code><a href="gc_hg38_1000kb.html">gc_hg38_50kb</a></code> <code><a href="gc_hg38_1000kb.html">gc_hg38_500kb</a></code> </p>
        </td>
        <td><p>GC content across the human genome build hg38/GRCh38</p></td>
      </tr><tr><td>
          <p><code><a href="getAnnotationDataFrame.html">getAnnotationDataFrame()</a></code> </p>
        </td>
        <td><p>This function generates a data frame to use for the annotation of heatmaps, when using groups</p></td>
      </tr><tr><td>
          <p><code><a href="getAnnotationDataFrameIndividual.html">getAnnotationDataFrameIndividual()</a></code> </p>
        </td>
        <td><p>This function generates a data frame to use for the annotation of heatmaps, using Individual sample names not group</p></td>
      </tr><tr><td>
          <p><code><a href="getBamCoveragePairedAndUnpairedR1.html">getBamCoveragePairedAndUnpairedR1()</a></code> </p>
        </td>
        <td><p>This function takes a bam file and gets the coverage over a set of regions. It keeps pairs or high quality R1s.</p></td>
      </tr><tr><td>
          <p><code><a href="getBetaMeans.html">getBetaMeans()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and calculates the mean of the nrpm and beta values over the windows given by GRanges.</p></td>
      </tr><tr><td>
          <p><code><a href="getBetaTable.html">getBetaTable()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and makes a table of beta values</p></td>
      </tr><tr><td>
          <p><code><a href="getCGPositions.html">getCGPositions()</a></code> </p>
        </td>
        <td><p>This function takes a GRanges object and calculates the enrichment scores.</p></td>
      </tr><tr><td>
          <p><code><a href="getCountTable.html">getCountTable()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and makes a table of nrpm values</p></td>
      </tr><tr><td>
          <p><code><a href="getDMRsData.html">getDMRsData()</a></code> </p>
        </td>
        <td><p>Extract the data table from all the contrasts</p></td>
      </tr><tr><td>
          <p><code><a href="getDataTable.html">getDataTable()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and generates a table of data containing data for each sample</p></td>
      </tr><tr><td>
          <p><code><a href="getGenomicFeatureDistribution.html">getGenomicFeatureDistribution()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and calculates some stats as to which genomic regions the reads lie.</p></td>
      </tr><tr><td>
          <p><code><a href="getNRPMTable.html">getNRPMTable()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and makes a table of nrpm values</p></td>
      </tr><tr><td>
          <p><code><a href="getNormalisedReadSum.html">getNormalisedReadSum()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and sums over nrpm within a GRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="getPCAwithBatch.html">getPCAwithBatch()</a></code> </p>
        </td>
        <td><p>This function edits the getPCA function from qsea to allow for batch effect correction</p></td>
      </tr><tr><td>
          <p><code><a href="getSampleQCSummary.html">getSampleQCSummary()</a></code> </p>
        </td>
        <td><p>This function returns the most important columns of a qseaSet with respect to quality controls</p></td>
      </tr><tr><td>
          <p><code><a href="getWindowNames.html">getWindowNames()</a></code> </p>
        </td>
        <td><p>This function returns the window names from a data frame with DMR windows</p></td>
      </tr><tr><td>
          <p><code><a href="hg19ToHg38.over.chain.html">hg19ToHg38.over.chain</a></code> </p>
        </td>
        <td><p>hg19tohg38 liftover chain</p></td>
      </tr><tr><td>
          <p><code><a href="hg38CpGIslands.html">hg38CpGIslands</a></code> </p>
        </td>
        <td><p>CpG islands for hg38.</p></td>
      </tr><tr><td>
          <p><code><a href="hg38UltraStableProbes.html">hg38UltraStableProbes</a></code> </p>
        </td>
        <td><p>CpG islands for hg38.</p></td>
      </tr><tr><td>
          <p><code><a href="hg38_450kArrayGR.html">hg38_450kArrayGR</a></code> </p>
        </td>
        <td><p>Infinium 450k probe locations for hg38.</p></td>
      </tr><tr><td>
          <p><code><a href="is.qseaSet.html">is.qseaSet()</a></code> </p>
        </td>
        <td><p>Check if an object is a qseaSet</p></td>
      </tr><tr><td>
          <p><code><a href="keepWindowsOverCutoff.html">keepWindowsOverCutoff()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and keeps the windows which have expression above a cutoff in a defined set of samples</p></td>
      </tr><tr><td>
          <p><code><a href="leftJoinQset.html">leftJoinQset()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and appends data to its sampleTable based on a call to dplyr::left_join</p></td>
      </tr><tr><td>
          <p><code><a href="left_join.qseaSet.html">left_join(<i>&lt;qseaSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>This function extends the dplyr function left_join to act on qseaSet sampleTable.</p></td>
      </tr><tr><td>
          <p><code><a href="liftOverhg19.html">liftOverhg19()</a></code> </p>
        </td>
        <td><p>This function takes a genomic ranges object or data frame, and lifts over from hg19 to hg38.</p></td>
      </tr><tr><td>
          <p><code><a href="makeAllContrasts.html">makeAllContrasts()</a></code> </p>
        </td>
        <td><p>Make a full set of contrasts</p></td>
      </tr><tr><td>
          <p><code><a href="makeQset.html">makeQset()</a></code> </p>
        </td>
        <td><p>Make the initial qseaSet object from a set of samples</p></td>
      </tr><tr><td>
          <p><code><a href="makeTransposedTable.html">makeTransposedTable()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and makes a wide transposed table, suitable for using for machine learning</p></td>
      </tr><tr><td>
          <p><code><a href="map_hg38_1000kb.html">map_hg38_1000kb</a></code> <code><a href="map_hg38_1000kb.html">map_hg38_10kb</a></code> <code><a href="map_hg38_1000kb.html">map_hg38_50kb</a></code> <code><a href="map_hg38_1000kb.html">map_hg38_500kb</a></code> </p>
        </td>
        <td><p>GC content across the human genome build hg38/GRCh38</p></td>
      </tr><tr><td>
          <p><code><a href="mergeQsetSamples.html">mergeQsetSamples()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and merges samples together into a single sample.</p></td>
      </tr><tr><td>
          <p><code><a href="mixArrayWithQset.html">mixArrayWithQset()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and makes a new sample by mixing two samples</p></td>
      </tr><tr><td>
          <p><code><a href="mixQsetSamples.html">mixQsetSamples()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and makes a new sample by mixing two samples</p></td>
      </tr><tr><td>
          <p><code><a href="mixThreeQsetSamples.html">mixThreeQsetSamples()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and makes a new sample by mixing three samples</p></td>
      </tr><tr><td>
          <p><code><a href="mutate.qseaSet.html">mutate(<i>&lt;qseaSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>This function extends the dplyr function mutate to act on qseaSet sampleTable.</p></td>
      </tr><tr><td>
          <p><code><a href="mutateQset.html">mutateQset()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and mutates its sampleTable based on a call to dplyr::mutate.</p></td>
      </tr><tr><td>
          <p><code><a href="pivotDMRsLonger.html">pivotDMRsLonger()</a></code> </p>
        </td>
        <td><p>This function transforms a DMR data frame from the wide format to a long format</p></td>
      </tr><tr><td>
          <p><code><a href="plotCNVheatmap.html">plotCNVheatmap()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and plots a heatmap of the calculated CNV</p></td>
      </tr><tr><td>
          <p><code><a href="plotCorrelationMatrix.html">plotCorrelationMatrix()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and plots a correlation matrix</p></td>
      </tr><tr><td>
          <p><code><a href="plotDMRUpSet.html">plotDMRUpSet()</a></code> </p>
        </td>
        <td><p>This function takes a DMRtable as output by calculateDMRs (possibly after any filtering), and generates an upset plot</p></td>
      </tr><tr><td>
          <p><code><a href="plotGRangesHeatmap.html">plotGRangesHeatmap()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and a GRanges object, and plots the expression across the gene as a heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotGeneHeatmap.html">plotGeneHeatmap()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and a gene, and plots the expression across the gene as a heatmap</p></td>
      </tr><tr><td>
          <p><code><a href="plotGenomicFeatureDistribution.html">plotGenomicFeatureDistribution()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and extracts the distribution of the genomic features of all the windows over a cutoff, and plots it</p></td>
      </tr><tr><td>
          <p><code><a href="plotQseaPCA.html">plotQseaPCA()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and plots a PCA</p></td>
      </tr><tr><td>
          <p><code><a href="pull.qseaSet.html">pull(<i>&lt;qseaSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>This function extends the dplyr function pull to act on qseaSet sampleTable.</p></td>
      </tr><tr><td>
          <p><code><a href="pullQset.html">pullQset()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and pulls a column from its sampleTable based on a call to dplyr::pull</p></td>
      </tr><tr><td>
          <p><code><a href="qseaTableToChrGRanges.html">qseaTableToChrGRanges()</a></code> </p>
        </td>
        <td><p>Convert a table from makeTable into a GRanges object</p></td>
      </tr><tr><td>
          <p><code><a href="qseaWriteDMRsBeds.html">qseaWriteDMRsBeds()</a></code> </p>
        </td>
        <td><p>Write the qsea result to a set of bed file</p></td>
      </tr><tr><td>
          <p><code><a href="qseaWriteDMRsExcel.html">qseaWriteDMRsExcel()</a></code> </p>
        </td>
        <td><p>Write the qsea result to an Excel file</p></td>
      </tr><tr><td>
          <p><code><a href="relabelQset.html">relabelQset()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and renames the samples based on a column of the sampleTable.</p></td>
      </tr><tr><td>
          <p><code><a href="removeCNV.html">removeCNV()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and removes the CNV data.</p></td>
      </tr><tr><td>
          <p><code><a href="removeLibraryFactors.html">removeLibraryFactors()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and sets all the library factor columns to one, including in the sampleTable.</p></td>
      </tr><tr><td>
          <p><code><a href="removeWindowsOverCutoff.html">removeWindowsOverCutoff()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and filters out windows which have expression in a defined set of samples</p></td>
      </tr><tr><td>
          <p><code><a href="remove_almost_empty_cols.html">remove_almost_empty_cols()</a></code> </p>
        </td>
        <td><p>This function removes columns that are mostly NA (with some proportion)
Based off the janitor::remove_empty_cols function.</p></td>
      </tr><tr><td>
          <p><code><a href="renameQsetNames.html">renameQsetNames()</a></code> </p>
        </td>
        <td><p>Rename samples in a qseaSet.</p></td>
      </tr><tr><td>
          <p><code><a href="runHMMCopy.html">runHMMCopy()</a></code> </p>
        </td>
        <td><p>Function to call HMMcopy on the individual reads for a sample</p></td>
      </tr><tr><td>
          <p><code><a href="select.qseaSet.html">select(<i>&lt;qseaSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>This function extends the dplyr function select to act on qseaSet sampleTable. Can also be used to rename columns.</p></td>
      </tr><tr><td>
          <p><code><a href="selectQset.html">selectQset()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and selects columns from its sampleTable based on a call to dplyr::select</p></td>
      </tr><tr><td>
          <p><code><a href="sort.qseaSet.html">sort(<i>&lt;qseaSet&gt;</i>)</a></code> </p>
        </td>
        <td><p>This function extends the function sort to act on qseaSet, to reorder the names. It uses gtools::mixedsort to sort numbers correctly (i.e. 10 does not come before 2).</p></td>
      </tr><tr><td>
          <p><code><a href="sortQset.html">sortQset()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and sorts the samples in it, so they are in alphabetical order (with numbers increasing via mixedsort)</p></td>
      </tr><tr><td>
          <p><code><a href="subsetQset.html">subsetQset()</a></code> </p>
        </td>
        <td><p>Subset a qseaSet</p></td>
      </tr><tr><td>
          <p><code><a href="subsetWindowsOverBackground.html">subsetWindowsOverBackground()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and finds which windows have more reads than would be expected if the reads followed a Poisson distribution over the whole genome.</p></td>
      </tr><tr><td>
          <p><code><a href="summariseByGene.html">summariseByGene()</a></code> </p>
        </td>
        <td><p>This function summarises a data frame by gene (using annotateData)</p></td>
      </tr><tr><td>
          <p><code><a href="summariseDMRsByContrast.html">summariseDMRsByContrast()</a></code> </p>
        </td>
        <td><p>This function summarises the number of windows called as up/down in each comparison in a .
Internally transforms the data to a long format. Contrasts with no significant windows will be dropped!</p></td>
      </tr><tr><td>
          <p><code><a href="writeBigWigs.html">writeBigWigs()</a></code> </p>
        </td>
        <td><p>This function takes a qseaSet and writes individual bigWig files with the coverage over each window.</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Simon Pearce, Steven Hill.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.5.</p>
</div>

      </footer></div>

  


  

  </body></html>

