<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Make the initial qseaSet object from a set of samples — makeQset • mesa</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Make the initial qseaSet object from a set of samples — makeQset"><meta property="og:description" content="This function makes the initial qseaSet from a sampleTable that includes the file paths for the bam files and their metadata"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">mesa</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.2.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Make the initial qseaSet object from a set of samples</h1>
    
    <div class="hidden name"><code>makeQset.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function makes the initial qseaSet from a sampleTable that includes the file paths for the bam files and their metadata</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">makeQset</span><span class="op">(</span></span>
<span>  <span class="va">sampleTable</span>,</span>
<span>  BSgenome <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  chrSelect <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">22</span>,</span>
<span>  windowSize <span class="op">=</span> <span class="fl">300</span>,</span>
<span>  CNVwindowSize <span class="op">=</span> <span class="fl">1e+06</span>,</span>
<span>  fragmentType <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fragmentLength <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  fragmentSD <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  CNVmethod <span class="op">=</span> <span class="st">"HMMdefault"</span>,</span>
<span>  coverageMethod <span class="op">=</span> <span class="st">"PairedAndR1s"</span>,</span>
<span>  minMapQual <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  minInsertSize <span class="op">=</span> <span class="fl">70</span>,</span>
<span>  maxInsertSize <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>  minReferenceLength <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  badRegions <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  properPairsOnly <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>sampleTable</dt>
<dd><p>A data frame with the sample information to be passed to qsea.</p></dd>


<dt>BSgenome</dt>
<dd><p>A BSgenome string. See BSgenome::available.genomes() for options.</p></dd>


<dt>chrSelect</dt>
<dd><p>Which chromosomes to use (default 1:22).</p></dd>


<dt>windowSize</dt>
<dd><p>What window size (in bp) to use in the genome (default 300).</p></dd>


<dt>CNVwindowSize</dt>
<dd><p>What window size (in bp) to use in the calculation of CNV (default 1000000).</p></dd>


<dt>fragmentType</dt>
<dd><p>What type of procedure generated the library, Sheared or cfDNA (used to set the average fragment length/SD to a default for CBC)</p></dd>


<dt>fragmentLength</dt>
<dd><p>Average DNA fragment length. Can be set by fragmentType.</p></dd>


<dt>fragmentSD</dt>
<dd><p>Standard deviation of the DNA fragment lengths. Can be set by fragmentType.</p></dd>


<dt>CNVmethod</dt>
<dd><p>Which method to use for calculation of CNV. Options include "HMMdefault" (hmmcopy with default parameters) and "None"</p></dd>


<dt>coverageMethod</dt>
<dd><p>Whether to use custom methtools method for reading coverage in (set to PairedAndR1s), rather than qsea's (qseaPaired).</p></dd>


<dt>minMapQual</dt>
<dd><p>Minimum MAPQ score for a read to be kept. For PairedAndR1s, will keep if either R1 or R2 meet this cutoff in a properly paired read (if MQ tags are set in the bam file with samtools fixmate).</p></dd>


<dt>minInsertSize</dt>
<dd><p>For paired reads, only keep them if they are above a minimum length. Can be used for cfDNA size selection. Applies to Input samples as well as MeCap.</p></dd>


<dt>maxInsertSize</dt>
<dd><p>For paired reads, only keep them if they are below a maximum length. Can be used for cfDNA size selection. Applies to Input samples as well as MeCap.</p></dd>


<dt>minReferenceLength</dt>
<dd><p>A minimum distance on the genome to keep the read. bwa by default gives 19bp as minimum for a read, which is quite short.</p></dd>


<dt>badRegions</dt>
<dd><p>A GRanges object containing regions to filter out from the result.</p></dd>


<dt>properPairsOnly</dt>
<dd><p>Whether to only keep properly paired reads, or to keep high-quality (MAPQ 30+) unpaired R1s as well. Set to TRUE for size selection.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>A qseaSet object, containing all the information required.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Simon Pearce, Steven Hill.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.6.</p>
</div>

      </footer></div>

  


  

  </body></html>

